pycqed/instrument_drivers/meta_instrument/qubit_objects/mock_CCL_Transmon.py

```diff
@@ -60,164 +62,284 @@ def add_mock_params(self):
-            a = ma.Homodyne_Analysis(label=name)
-
-            if use_min:
-                f_res = a.min_frequency
+            freq = items[0]
+            if with_VNA:
+                VNA = self.instr_VNA.get_instr()
+                start_freq = freq - 10e6
+                stop_freq = freq + 10e6
+                name = 'VNA_Resonator_scan_' + str(round(freq/1e9, 4)) + 'GHz'
+                self.measure_with_VNA(VNA, start_freq, stop_freq, npts)
+            else:
+                freqs = np.arange(freq-5e6, freq+5e6, 0.1e6)
+                name = 'Resonator_Scan' + self.msmt_suffix
+                self.measure_heterodyne_spectroscopy(freqs=freqs,
+                                                     analyze=False)
+
+            a = ma.Homodyne_Analysis(label=name, qb_name=self.name)
+
+            dip = np.amin(a.data_y)
+            offset = a.fit_results.params['A'].value
+
+            if np.abs(dip/offset) > 0.5:
+                print('Removed candidate {} ({:.3f} GHz): Not a resonator'
+                      .format(resonator, freq/1e9))
+                delkeys.append(resonator)
+            elif np.isnan(a.fit_results.params['Qc'].stderr):
+                print('Removed candidate {} ({:.3f} GHz): Not a resonator'
+                      .format(resonator, freq/1e9))
+                delkeys.append(resonator)
             else:
-                f_res = a.fit_results.params['f0'].value*1e9
-            self.res_dict[resonator][1] = f_res
+                if use_min:
+                    f_res = a.min_frequency
+                else:
+                    f_res = a.fit_results.params['f0'].value*1e9
+
+                # Check if not a duplicate
+                i = int(resonator)
+                if i > 0:
+                    prev_freq = self.res_dict[str(i-1)][0]
+                    if np.abs(prev_freq - f_res) < 10e6:
+                        delkeys.append(resonator)
+                        print('Removed candidate: ' + resonator + ' (' +
+                              str(round(f_res/1e9, 3)) + ' GHz): Duplicate')
+                    else:
+                        self.res_dict[resonator][0] = f_res
+                        print("Added resonator " + resonator + ' (' +
+                              str(round(f_res/1e9, 3)) + ' GHz)')
+
+        for delkey in delkeys:
+            self.res_dict.pop(delkey)
+
+        # Rearrange dictionary to start from 0 again:
+        i = 0
+        newdict = {}
+        for resonator, items in self.res_dict.items():
+            newdict[str(i)] = items
+            i += 1
 
-        # if f_res > max(freqs) or f_res < max(freqs):
-        #     logging.warning(
-        #         'extracted frequency outside range of scan. Not updating')
-        if update:
-            self.freq_res(f_res)
-            self.ro_freq(f_res)
+        self.res_dict = newdict
         return True
 
-    def find_test_resonators_VNA(self):
+    def find_test_resonators(self, with_VNA=None):
```

Comment:
Should not be in qubit object, but I think we have already discussed this. May be a good idea to have a planned meeting on this subject.